components/wallet/lnd.js

```diff
@@ -0,0 +1,110 @@
+import React from 'react'
+
+import { ensureB64 } from '@/lib/format'
+import { datePivot } from '@/lib/time'
+import { LNDAutowithdrawSchema } from '@/lib/validate'
+
+export const name = 'lnd'
+
+export const fields = [
+  {
+    name: 'socket',
+    label: 'grpc host and port',
+    type: 'text',
+    placeholder: '55.5.555.55:10001',
+    hint: 'tor or clearnet',
+    clear: true
+  },
+  {
+    name: 'macaroon',
+    label: 'invoice macaroon',
+    help: {
+      label: 'privacy tip',
+      text: 'We accept a prebaked ***invoice.macaroon*** for your convenience. To gain better privacy, generate a new macaroon as follows:\n\n```lncli bakemacaroon invoices:write invoices:read```'
+    },
+    type: 'text',
+    placeholder: 'AgEDbG5kAlgDChCn7YgfWX7uTkQQgXZ2uahNEgEwGhYKB2FkZHJlc3MSBHJlYWQSBXdyaXRlGhcKCGludm9pY2VzEgRyZWFkEgV3cml0ZRoPCgdvbmNoYWluEgRyZWFkAAAGIJkMBrrDV0npU90JV0TGNJPrqUD8m2QYoTDjolaL6eBs',
+    hint: 'hex or base64 encoded',
+    clear: true
+  },
+  {
+    name: 'cert',
+    label: 'cert',
+    type: 'text',
+    placeholder: 'LS0tLS1CRUdJTiBDRVJUSUZJQ0FURS0tLS0tCk1JSUNNVENDQWRpZ0F3SUJBZ0lRSHBFdFdrcGJwZHV4RVF2eVBPc3NWVEFLQmdncWhrak9QUVFEQWpBdk1SOHcKSFFZRFZRUUtFeFpzYm1RZ1lYVjBiMmRsYm1WeVlYUmxaQ0JqWlhKME1Rd3dDZ1lEVlFRREV3TmliMkl3SGhjTgpNalF3TVRBM01qQXhORE0wV2hjTk1qVXdNekF6TWpBeE5ETTBXakF2TVI4d0hRWURWUVFLRXhac2JtUWdZWFYwCmIyZGxibVZ5WVhSbFpDQmpaWEowTVF3d0NnWURWUVFERXdOaWIySXdXVEFUQmdjcWhrak9QUUlCQmdncWhrak8KUFFNQkJ3TkNBQVJUS3NMVk5oZnhqb1FLVDlkVVdDbzUzSmQwTnBuL1BtYi9LUE02M1JxbU52dFYvdFk4NjJJZwpSbE41cmNHRnBEajhUeFc2OVhIK0pTcHpjWDdlN3N0Um80SFZNSUhTTUE0R0ExVWREd0VCL3dRRUF3SUNwREFUCkJnTlZIU1VFRERBS0JnZ3JCZ0VGQlFjREFUQVBCZ05WSFJNQkFmOEVCVEFEQVFIL01CMEdBMVVkRGdRV0JCVDAKMnh3V25GeHRUNzI0MWxwZlNoNm9FWi9UMWpCN0JnTlZIUkVFZERCeWdnTmliMktDQ1d4dlkyRnNhRzl6ZElJRApZbTlpZ2d4d2IyeGhjaTF1TVMxaWIyS0NGR2h2YzNRdVpHOWphMlZ5TG1sdWRHVnlibUZzZ2dSMWJtbDRnZ3AxCmJtbDRjR0ZqYTJWMGdnZGlkV1pqYjI1dWh3Ui9BQUFCaHhBQUFBQUFBQUFBQUFBQUFBQUFBQUFCaHdTc0VnQUQKTUFvR0NDcUdTTTQ5QkFNQ0EwY0FNRVFDSUEwUTlkRXdoNXpPRnpwL3hYeHNpemh5SkxNVG5yazU1VWx1NHJPRwo4WW52QWlBVGt4U3p3Y3hZZnFscGx0UlNIbmd0NUJFcDBzcXlHL05nenBzb2pmMGNqQT09Ci0tLS0tRU5EIENFUlRJRklDQVRFLS0tLS0K',
+    // worker does not support JSX syntax
+    optional: React.createElement(
+      React.Fragment,
+      {},
+      'optional if from ',
+      React.createElement('a', { href: 'https://en.wikipedia.org/wiki/Certificate_authority', target: '_blank', rel: 'noreferrer' }, 'CA'),
+      ' (e.g. voltage)'),
+    hint: 'hex or base64 encoded',
+    clear: true
+  }
+]
+
+export const card = {
+  title: 'LND',
+  subtitle: 'autowithdraw to your Lightning Labs node',
+  badges: ['receive only', 'non-custodial']
+}
+
+export const schema = LNDAutowithdrawSchema
+
+export const server = {
+  walletType: 'LND',
+  walletField: 'walletLND',
+  resolverName: 'upsertWalletLND',
```

Comment:
Mhh, no, I think you are right. When I wrote this code, I wasn't sure if injecting GraphQL typedefs and resolver was already too much magic so I at least kept the names of what gets injected explicit, lol

Reply:
Mhh, I actually think `walletType` and `walletField` are good reminders that database changes are required for receiving wallets. `walletType` must match a `WalletType` enum value and `walletField`  must match a column in the `Wallet` table (as documented in _example.js). I could assume that they will always be called in this way but I think it will lead to "oh, you can't name your column / enum value that way". So it's better to be explicit here imo.

The value for `resolverName` however can be set to anything. So I will add code that always uses `upsertWallet${module.name}` (removing any characters that might be illegal in this context).

Reply:
I decided to use `walletField` in f01ce79a since it's semantically closer to what the resolver is about